src/gretel_trainer/models.py

```diff
@@ -0,0 +1,81 @@
+"""Model specific parameters and definitions"""
+
+from collections import namedtuple
+from enum import Enum
+
+
+class BaseConfig:
+    """This class should not be used directly, model-specific configs should be
+    derived from this class
+    """
+
+    def __init__(
+        self,
+        config_file: str,
+        default_rows: int,
+        max_rows: int,
+        default_header_clusters: int,
+        max_header_clusters: int,
+    ):
+        self.config_file = config_file
+        self.default_rows = default_rows
+        self.max_rows = max_rows
+        self.default_header_clusters = default_header_clusters
+        self.max_header_clusters = max_header_clusters
+
+
+class GretelLSTMConfig(BaseConfig):
+    def __init__(self):
+        super().__init__(
+            config_file="synthetics/default",
+            default_rows=50000,
+            max_rows=1000000,
+            default_header_clusters=20,
+            max_header_clusters=30,
+        )
+
+
+class GretelCTGANConfig(BaseConfig):
+    def __init__(self):
+        super().__init__(
+            config_file="synthetics/high-dimensionality",
+            default_rows=50000,
+            max_rows=1000000,
+            default_header_clusters=500,
+            max_header_clusters=1000,
+        )
+
+
+class ExtendedEnum(Enum):
+    """Utility class for Model enum"""
+
+    @classmethod
+    def get_types(cls):
+        return list(map(lambda c: c.name, cls))
+
+    @classmethod
+    def get_config(cls, model: str) -> str:
+        return cls[model].config.config_file
+
+    @classmethod
+    def get_default_rows(cls, model: str) -> int:
+        return cls[model].config.default_rows
+
+    @classmethod
+    def get_max_rows(cls, model: str) -> int:
+        return cls[model].config.max_rows
+
+    @classmethod
+    def get_default_header_clusters(cls, model: str) -> int:
+        return cls[model].config.default_header_clusters
+
+    @classmethod
+    def get_max_header_clusters(cls, model: str) -> int:
+        return cls[model].config.max_header_clusters
+
+
+class Model(namedtuple("Model", "config"), ExtendedEnum):
```

Comment:
I'm not sure I understand the purpose of this enum as well as the wrapper around it with the extended enum above. Couldn't we just use instances of `BaseConfig` as a param to the `Trainer` class?

Instead of `model_type: str` as a param to `Trainer` could we just pass in a `model_config: BaseConfig`? This is similar to what we do in `gretel-synthetics` and just use a builder pattern.

Also on the `Trainer` I'm not sure why `max_rows` and `max_header_clusters` need to params, it seems like they could just be only defined on the config instance?

So trainer could be something like:

`trainer = Trainer(model_config=GretelLSTMConfig()` then you don't need all the getters from the enum you can just access the attributes directly like `trainer.model_config.max_rows`

Reply:
I didn't test this, but this would give you a more self-contained config that is self-validating. Not sure we need to swap to something like `pydantic` just yet. You can define your constraints i.e. `_max_rows` as actual _class_ attributes because numbers and strings are immutable. This way you don't have to clutter up the constructors with constraints the user shouldn't really need to think about.

Also, on the concrete constructors, you will probably want to have your default values declared on the actual `__init__` because as it stands now there's no way to change these values. It looks like you were actually trying to replicate these values into the `Trainer` as attributes, but I'd suggest just keeping everything contained within the config class instance.

```python
class _BaseConfig:
    """This class should not be used directly, model-specific configs should be
    derived from this class
    """

    _max_header_clusters: int
    """This should be overridden on conrete classes"""

    # Default values for all configs, may be overridden if needed
    _max_rows: int = 1000000

    # Should be set by concrete constructors
    config_file: str
    default_rows: int
    default_header_clusters: int

    def __init__(
        self,
        config_file: str,
        default_rows: int,
        default_header_clusters: int,
    ):
        self.config_file = config_file
        self.default_rows = default_rows
        self.default_header_clusters = default_header_clusters

        self.validate()

    def validate(self):
        if self.default_rows > self._max_rows:
            raise ValueError("too many rows")

        if self.default_header_clusters > self._max_header_clusters:
            raise ValueError("too many header clusters")


class GretelLSTMConfig(_BaseConfig):

    _max_header_clusters: int = 30

    def __init__(
        self,
        config_file="synthetics/default",
        default_rows=50000,
        default_header_clusters=20,
    ):
        super().__init__(
            config_file=config_file,
            default_rows=default_rows,
            default_header_clusters=default_header_clusters,
        )


class GretelCTGANConfig(_BaseConfig):

    _max_header_clusters: int = 1000

    def __init__(
        self,
        config_file="synthetics/high-dimensionality",
        default_rows=50000,
        default_header_clusters=500,
    ):
        super().__init__(
            config_file=config_file,
            default_rows=default_rows,
            default_header_clusters=default_header_clusters,
        )
```

Reply:
I used the Enum to let the code enumerate the available model types and check that whatever the user was asking for was a valid type. But agreed it's a little extra and we can add this to the class functions anyways. Let's go with the changes above 👍